### PR TITLE
fix: change dynamic html string w/CSP safe code to fix scroll, fix #914

### LINKED ISSUE #914: Horizontal scrolling causes cells to render in wrong location
Horizontal scrolling a particularly wide SlickGrid causes cell content to be mis-rendered.

This can be seen in [example 1](https://6pac.github.io/SlickGrid/examples/example1-simple.html) by simply dragging the columns wider and then scrolling horizontally:

Example of horizontal scrolling issue
(Recorded in Firefox Dev Edition 120.0b9; same occurs in Edge 119.0.2151.58)

This seems to have been introduced in 5.5.1 as I can't reproduce with 5.4.2.

### PR DESCRIPTION
**Pull request by JesperJakobsenCIM**
This is not ready to get merged, its just to have a place to discuss for now

### CODE REVIEW

**Comment by JesperJakobsenCIM** [comment]
Okay I'm experiencing weird things, if you scroll from left to right fast, the data shows up and then disappears shortly after rendering (seems to be column auto correct)
https://streamable.com/2ol752 from 7 seconds and forward

Do you have any ideas where the code for this logic exists? since my guess would be its a change in structure from old to new and there are some code that still expects the old style

**Comment by ghiscoding** [comment]
oh that's great, never taught of looking at the DOM element it creates, the new Cypress test I added last week should cover this use case pretty well, I made it scroll slowly since it showed more often while scrolling slowly. Out of curiosity, could I see the code you changed since last time, I'd like to see where the div in a div happened. 

is it ready to be merged then? I think it looks good for my point of view since all Cypress E2E tests are passing, including the new test I added. Perhaps the only other thing I could would be to test a little more locally.

File: src/slick.grid.ts [hunk 1]
```diff
@@ -3923,27 +3920,25 @@ export class SlickGrid<TData = any, C extends Column<TData> = Column<TData>, O e
     if ((formatterResult as FormatterResultObject)?.addClasses) {
       addlCssClasses += (addlCssClasses ? ' ' : '') + (formatterResult as FormatterResultObject).addClasses;
     }
-    const toolTip = (formatterResult as FormatterResultObject)?.toolTip ? `title="${(formatterResult as FormatterResultObject).toolTip}"` : '';
 
-    let customAttrStr = '';
+    const toolTipText = (formatterResult as FormatterResultObject)?.toolTip ? `${(formatterResult as FormatterResultObject).toolTip}` : '';
+    const cellDiv = document.createElement('div');
+    cellDiv.className = cellCss + (addlCssClasses ? ' ' + addlCssClasses : '');
+    cellDiv.setAttribute('title', toolTipText);
```
Comment (ghiscoding) [inline src/slick.grid.ts:3927]: you could also use the `createDomElement` util, but in this case you would only save 1 line because util cannot deal with attributes, so you still need to call `setAttribute` separately. So anyway you could skip the use of the util for this one but I would nonetheless prefer to use string literal (with backtick ```) since I find it more readable.
```ts
cellDiv.className = `${cellCss}${addlCssClasses ? ' ' + addlCssClasses : ''}`);
```
maybe another way would be to use the string `trim()` function, that should remove the 2 spaces it adds though it's probably better to verify the output in the browser.
```ts
cellDiv.className = `${cellCss} ${addlCssClasses}`).trim();
```

**Comment by JesperJakobsenCIM** [comment]
This code is ready to get merged

**Comment by ghiscoding** [comment]
Thanks I tried it locally and see no more issues, good job 👍🏻 
I'll go ahead and merge, expect a new release by the weekend

**Comment by ghiscoding** [comment]
@JesperJakobsenCIM so while I was giving this a try in Slickgrid-Universal (my lib based on slickgrid), I found that I had an issue in my lib after applying your changes, the issue was around the lines shown below. I thought it might interesting for you to know since it's CSP related

https://github.com/6pac/SlickGrid/blob/1a93a64c1021cb74bdeee1859bcbe2440eb26ee2/src/slick.grid.ts#L3934-L3937

mainly the `applyHtmlCode` is a new method I've added in PR #894 which allow us to code a Formatter with native HTML element or keep using HTML strings and the `applyHtmlCode` will either `.appendChild()` when it's a native element or when it's a string it will sanitize it and then use `innerHTML`.... Anyway, the issue I had was with grid that have Grouping, the row that is a Group has an extra `level="1"` representing the depth level of the group (0 being at the root level, see image below). So in my case, I use DOMPurify as sanitizer and `applyHtmlCode` passes HTML string to the sanitizer when it is a string and what took me some time to realize is that DOMPurify strips any attributes that it doesn't know unless we tell them it's allowed, basically I had to change the DOMPurify to use these options `DOMPurify.sanitize(val || '', { ADD_ATTR: ['level'], RETURN_TRUSTED_TYPE: true })` and that fixed my issue. 

So long story short, the DOMPurify issue brought up another change that should be addressed eventually, I'll probably work at it in the next couple days before the release, there are still many areas of the code that uses Formatters with HTML strings and that could cause other indirect issues like I had, converting the code below to be native HTML element would help with CSP compliance (which is what I will do)... or in other words, there might be other areas of SlickGrid core or plugins that need refactoring to be fully CSP compliant and there are still areas of the code that still require `innerHTML`, if you find others then please open a PR. 

https://github.com/6pac/SlickGrid/blob/1a93a64c1021cb74bdeee1859bcbe2440eb26ee2/src/slick.groupitemmetadataprovider.ts#L62-L78

image

**Comment by ghiscoding** [comment]
@JesperJakobsenCIM as per what I wrote in previous paragraph, I created PR #925 to convert Group Formatter to native HTML elements, I also used `DocumentFragment` which I didn't know that I could use to basically create an empty shell since the previous html string had 2 span inside and I didn't want to create an empty div container, I saw fragment in the past but didn't know I could use it this way, anyway it's a fun fact :) ... you could comment on PR #925 if you want, I'll probably merge and release later tonight since I have other fixes that my team is waiting for and I prefer to do it through an official release.
